### ISSUE TITLE
Rewrite the Admin UI in React

### ISSUE THREAD

**Issue by JedWatson** [keystonejs/keystone#503]
### Some Background

The KeystoneJS Admin UI was originally built as a 'simple' web app, based on old-school form generation, submission and rendering techniques combined with jQuery for client-side UI enhancements.

This decision has helped us develop reusable processing techniques that are easily applied to server-driven projects (from contact form processing on websites to complete applications and APIs).

It's also been limiting though, and ultimately we want the Admin UI to be a rich client-side app. Now that the server side patterns are mature, it's time to focus on that.

Because we've known for a while that the Admin UI will be rewritten, this has unfortunately blocked several enhancements we've been keen to build, and which have often been requested by our community, including a plugin architecture, custom field types, and richer Admin UI functionality.

Primarily, the block has been that if we open it up to customisation then change the whole architecture, it'll create a backwards compatibility nightmare for anyone who's invested in the old system. So this is now the next big thing, that will lead to other big things, and I'm really excited to be digging into it.

### Why React?

I've spent a long time trying to decide the best fit for Keystone's Admin UI, and have assessed a lot of frameworks including Angular, Ember, Polymer, Backbone and more before settling on React.

Polymer was the favourite for a long time - because web components are the future, and play a very important role in our long term plans for Keystone - but unfortunately it's just not mature enough.

React, on the other hand, has excellent support, a robust routing solution, is very fast, and does a great job of providing many of the benefits web components will bring, today.

It also plays very nicely with Browserify, which combined with a grunt workflow, provides a great development environment, and makes sense for the way we expect to implement custom field types and extensions.

### The Plan

We don't expect this to be a massive, disruptive rebuild - it doesn't need to be. The first step will be to selectively replace server-generated code with React components that still behave the way the current Admin UI does, with a form-submission / processing / rendering workflow.

Once that is done, many of the blocks around allowing extensible field types and Admin UI customisation will be removed.

The next step will be to transition fully to a react app that will run client-side against API endpoints. In the same way we made the `UpdateHandler` and other tools work in a way that dramatically improves productivity when developing server-side processing workflows, we'll use this transition to develop great client-side app workflows that are reusable in projects. Very little will be specific to the Admin UI itself.

For the field types (and forms, and list UI) our intention is to develop reusable components that can be dropped into Keystone projects and reused outside of the Admin UI, providing rich functionality regardless of whether the server or client is doing the processing.

Some things will certainly need a rich client interface, but server-side HTML generation and processing is still important on the web, and hopefully we can create a solution that facilitates the holy grail of sharing code between the client and the server without being locked into either.

### Can you help?

There is still a lot to be worked out around the best way to manage application state, extensibility and component structure. There will also be quite a bit of work involved in transitioning all the current UI elements to React components.

If you have experience working with React and would be happy to help, please let us know!

**Comment by cameronjroe** [comment]
Awesome, I've been using React a lot lately. Would be happy to contribute to the migration. Is there a react branch somewhere?

**Comment by BerkeleyTrue** [comment]
I believe for the most part this is complete. Admin in 0.3.0 is completely in react. @JedWatson Should this issue be closed?

**Comment by aksonov** [comment]
Great! Any plans to use React at other parts? Ot it is possible already?

**Comment by BerkeleyTrue** [comment]
If you are asking about using Keystone with a react app, yes it is possible.

**Comment by mgan59** [comment]
@aksonov just modify your keystone's express to use a ReactJS template/view engine.

**Comment by morenoh149** [comment]
closing 0.3.x uses react for the adminUI

**Comment by JedWatson** [comment]
There's still a lot to do. Will explain shortly.

**Comment by JedWatson** [comment]
We've rewritten the Admin UI Forms in React and migrated the field types. This was a pretty major undertaking as it involved a complete rewrite of the field UI components.

That's really a prerequisite task though. We're getting closer to having the Admin UI in React but there's a lot we can now do. The ultimate goal is to have made the Admin UI a SPA.

I'm going to open some more granular issues for these as I spec them, but generally the roadmap looks like this:

- [ ] Rebuild List Filters / Sorting and Header UI in React
- [ ] Rebuild List (table) in React w/ Column component per Field type
- [ ] Use API to load List view entirely client-side (probably our first implementation of react-router) and move the rest of the UI (sorting, deleting, pagination, etc) to React
- [ ] Rebuild the Home screen and Header / Navigation Chrome in React
- [ ] Expose API endpoints for entirely client-side interactions (create / update items, etc)
- [ ] Use API endpoints for Item actions (create / save, etc) instead of submitting POST events
- [ ] Bring the whole thing together using react-router into a single page app

While this is being done, we'll start being able to open some parts of the Admin UI up for customisation, but I'm still planning to be conservative with it until we're confident with an API we're happy to support long-term.

I also want to flag that a large portion of the background work here is getting _all_ the field types under test - when we switch from POST to AJAX requests, I want to make sure that we have a way to ensure POSTing to all field types (e.g. for use via the UpdateHandler) continues to work as a first class use-case, even though it won't be used by the Admin UI anymore. @sebmck is generally leading this charge.

Once this roadmap is complete, it's going to be awesome. And then we can close this issue :)

In the meantime if anyone's interested in taking on part of this work it may not be obvious where or how to jump in, but there's lots we could use help with so reach out to me in our Slack (and if you're not in our Slack ping me, I'll add you)

**Comment by dandigangi** [comment]
Is there an update on where this rewrite stands?

**Comment by eshiferax** [comment]
^yeah, also wondering myself

**Comment by BerkeleyTrue** [comment]
The admin panel is written in react as of 0.3.0. This issue should be closed.

**Comment by JedWatson** [comment]
Stop trying to close this issue before it's done :P

The production release is still in the state I described above, with a mix of jade, server-side generation and React components.

Most of the items I listed above are _mostly_ done, - if you're interested in seeing the progress, check out the `elemental-integration` branch. We've created another project, [Elemental UI](http://elemental-ui.com) which we are using to develop the standard React form components for Keystone, which will also be available for use in projects outside the Admin UI.

I expect to release 0.4 within the next few weeks, having completed this transition.

**Comment by BerkeleyTrue** [comment]
IMHO It seems like that should really be a separate issue. Maybe something along the lines of 'elemental ui transition' Especially considering how much of a big deal the 0.3.0 release was. It also helps give a sense that the project is moving and not stalling on issues from years ago.







---

_<small>#AlwaysUseTwoSpaces ;)</small>_

**Comment by kfancy** [comment]
Hi @JedWatson et al, I'm late to the party but very interested and confident to start up a couple of large-scale projects with Keystone... and am a big fan of ReactJS. Very happy that you guys made that decision over the two-way data binding approach of others.

Not exactly sure how I can contribute (where to start, really), but I'm happy to if there's still some need.

**Comment by cameronjroe** [comment]
Any thoughts on using `react-router` for an isomorphic routing solution to keystone?

**Comment by morenoh149** [comment]
@kfancy there are issues labeled [easy-for-beginners](https://github.com/keystonejs/keystone/issues?q=is%3Aopen+is%3Aissue+label%3A%22easy+for+beginners%22) where we provide guidance

**Comment by kfancy** [comment]
I'm actively developing a couple of projects with react-router for exactly that behavior, I will be able to give some more in depth information on that topic soon but at minimum I'd say it's a solid package to work with.

Also, I'd recommend a pure flux implementation instead of using one of the myriad of "wrappers" (Fluxxor, etc) that sit on top of it. Flux is simply a philosophy at this point, and an easy one once you wrap your head around it... so far for me it plays well with ajax as well as websocket asynchronous communication.

**Comment by kfancy** [comment]
@morenoh149 what about open issues or subprojects working with react components?

**Comment by SOSANA** [comment]
This would be epic. Love react.js. woooooooohooo!

**Comment by cameronjroe** [comment]
@kfancy Yeah react-router would allow for route configuration of keystone isomorphically. Percolate has a nice [example](https://github.com/percolatestudio/percolatestudio.com/blob/master/app/server.js) just using `server.use` from express to run the routes from react-router.

**Comment by BerkeleyTrue** [comment]
@JedWatson I would definitely wait for React Router 1.0.0 (currently in beta) release. It has a fantastic new api and a scalable async routes config api that would work great for Keystones admin panel.

**Comment by grabbou** [comment]
Personally I think the `admin` UI should be moved into another package and required as a middleware. Would've simplified things a lot and also allowed further customisations w/o touching the core.

**Comment by BerkeleyTrue** [comment]
The admin panel is the CMS you can't have a CMS without it. If you remove it then keystone just becomes mongoose...

**Comment by grabbou** [comment]
I would argue with that, take a look at Django project & Django admin, which is an opt-in and it's not required for the project to work. For example, I can easily bootstrap REST api and use admin panel for quick dev scaffolding. 

Anyway, separating 'admin' to a different repo is just small improvement to make dev better (imo) - for end users it will be the same w/o any breaking changes or additional steps to take.

**Comment by JedWatson** [comment]
Separating the Admin UI from Keystone is part of a longer-term vision I've got for it to become more modular, at the moment it's not so simple as the field types are (unfortunately) quite tightly-coupled.

There's a lot of refactoring going around the API and express app that's probably going unnoticed that's preparing it for this.

We'll launch the React / Elemental version first because there will be quite a few breaking changes (to internals that I know are being used externally) but I expect this vision will be realised towards the end of the year.

As a brief overview:

Keystone is really several concepts that work together:

- Rich schema definition (currently implemented as a Mongoose functionality superset)
- Additional functionality (e.g UpdateHandler) on top of the Schemas
- Express app configuration (appropriate for most data-driven websites and apps)
- Admin UI generates from the Schemas

Basically the core is the schemas (i.e `keystone.List`) and the rich field types. My plan is to make them _much_ more portable and separate the functionality from its current tight coupling in the rest of the system (and rebuilding the Admin UI in React is the biggest step towards this). Bringing some of the schema / field type functionality to the client-side will be a big advantage as well.

Eventually each module will be able to be separated into its own package, and `keystone` will continue to bring them all together for the common use case (based on the principle "make the 80% easy, the 20% possible").

**Comment by lennerd** [comment]
I really would love to use Keystone for my next client project. Is there any chance to get an overview of your progress with this issue as it seems to block many other smaller issues (multi delete items in a list, sub collections, etc.)? Maybe some kind of roadmap issue I was missing while searching for something where the discussion takes place?

**Comment by niallobrien** [comment]
May I suggest using Reflux instead of Flux, simply to reduce the complexity and lower the barrier to entry? Thanks.

**Comment by kfancy** [comment]
First I'd have to ask how it lowers the barrier to entry? Flux is not a
hard architecture to understand, and once you put a wrapper in the way,
you're stuck with their implementation of the flux architecture. Upgrades?
Maybe they don't keep up with them.

My personal opinion, vanilla JS is always better unless there's an explicit
reason to use Yet Another Library.

**Comment by niallobrien** [comment]
In my opinion, the whole notion of using a dispatcher is a little unnecessary, Reflux solves this. It's like comparing Angular's ngRoute with UI-Router: the 3rd-party offering is just far superior (UI-Router).
It would also reduce the amount of code required.

**Comment by JedWatson** [comment]
I am currently building it using a simple data management pattern implemented with my [store-prototype](https://github.com/jedwatson/store-prototype) module. I view Flux as more of a "concept" than a "framework" and usually implement whatever level of it I feel is appropriate in different projects.

Implementation in progress for the List view is here, if you want a preview: https://github.com/keystonejs/keystone/blob/elemental-integration/admin/src/stores/CurrentListStore.js

If we do go with a more complete Flux architecture down the track, I expect it would be [redux](http://rackt.github.io/redux/) by @gaearon, he's doing amazing work on it.

Feedback on the current approach and state of the implementation in the `elemental-integration` branch is, of course, welcome, and if anybody wants to step up and help me with it that would be great. Get in touch on the Slack and I'll help you work out where to get started.

**Comment by dandigangi** [comment]
+1 Jed

Flux is an architectural pattern. If you're going to suggest it, provide the technical gains, not that it is just easier to use.

**Comment by dandigangi** [comment]
Suggest Reflux* sorry can't edit from mobile.

**Comment by gaearon** [comment]
>May I suggest using Reflux instead of Flux, simply to reduce the complexity and lower the barrier to entry?

Be aware that Reflux gives up most benefits of Flux attempting to be less verbose. [This](https://medium.com/@dan_abramov/the-evolution-of-flux-frameworks-6c16ad26bb31) might give you some idea.

**Comment by cameronjroe** [comment]
Hey Jed, just updated some of the API docs. Would love to join the Slack group!

**Comment by yeso126** [comment]
Redux?

**Comment by swagata-daffodil** [comment]
Hi JedWatson,
I have  following react_select items from drop-down as shown below plz let me know how can i make it draggable  for ordering content blocks items- 

![mudaimg](https://cloud.githubusercontent.com/assets/16575641/12291797/699f98f8-ba0f-11e5-8cad-f4de9ca3e0c7.PNG)

Please suggest I am using keystonejs where it is using react for admin ui ,so I need to drag and put blocks item accordingly.

**Comment by creynders** [comment]
@mxstbr is Chuck Norris'ing this one at #2566
